python/tank/util/metrics_cache.py

```diff
@@ -0,0 +1,93 @@
+# Copyright (c) 2023 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+"""
+The Metrics logger thread is only launched once an engine is initialized.
+But we might want to log metrics event before Toolkit is fully initialized.
+
+For instance, when starting ShotGrid Desktop and authenticating, there is no
+engine set yet.
+
+This module provide a cache that can be filled prior to Toolkit initilization.
+Then, once the engine is set and the metric thread starts, the cache is consumed
+and send the items to the Metric instance.
+
+Unfortunatly, this cache can not live in a singleton in this module because,
+somehow, tk-core is fully unloaded off the python process between authentication
+and launch of an engine (desktopstartup).
+Instead, we store the records in the environment as JSON text
+"""
+
+import hashlib
+import json
+import os
+
+from . import metrics
+from .. import platform
+
+from .. import LogManager
+
+logger = LogManager.get_logger(__name__)
+
+
+def log(*args, **kwargs):
+    """
+    Same prototype as the log method from the metrics module.
+    """
+
+    if platform.current_engine():
+        # We don't need to cache in this situation, Let's simply forward the
+        # order
+        return metrics.EventMetric.log(*args, **kwargs)
+
+    try:
+        cached_data = json.dumps((args, kwargs))
+    except TypeError:
+        logger.debug("Unable to cache metric. Can not be JSON encoded")
+        return
+
+    cache_key = "sgtk_metric_cache_{cid}".format(
+        cid=hashlib.sha1(cached_data.encode()).hexdigest()[:16],
+    )
+
+    os.environ[cache_key] = cached_data
+
+
+def consume():
+    """
+    Iterate the environment to find all the cached metrics
+    """
+
+    for cache_key in os.environ:
+        if not cache_key.startswith("sgtk_metric_cache_"):
+            continue
+
+        try:
+            cached_data = os.environ.pop(cache_key)
+        except KeyError:
+            # Might be due to parallel access
+            continue
+
+        try:
+            data = json.loads(cached_data)
+        except json.JSONDecodeError:
+            logger.debug("Unable to decode cached metric")
+            continue
+
+        if not isinstance(data, list) or len(data) != 2:
+            logger.debug("Invalid cached metric format")
+            continue
+
+        (args, kwargs) = data
+        if not isinstance(args, list) or not isinstance(kwargs, dict):
+            logger.debug("Invalid cached metric format")
+            continue
+
+        metrics.EventMetric.log(*args, **kwargs)
```

Comment:
I'm not 100% sure how EventMetric is working but It requires two str positional arguments first: `group`, `name`. Does this unpacking guarantees this?

Reply:
I think the answer to this question is there: https://www.geeksforgeeks.org/packing-and-unpacking-arguments-in-python/

This is one think about Python that I love :)

Reply:
I know 😅 but since we don't have type annotations, I'd like to provide a stronger explicit declaration. Anyway this is not a blocker as it seems `consume` is used just for this logic.